git_machete/git_operations.py

```diff
@@ -804,10 +804,53 @@ def is_equivalent_tree_reachable(
         )
 
         result = earlier_tree_hash in intermediate_tree_hashes
-        debug(f"result = {result}")
+        debug(f"intermediate tree hashes result = {result}")
+
+        if result:
+            self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = result
+            return result
+
+        # Let's try another way, a little more complex but takes into account the possibility
+        # that there were other commits between the fork point of the two branches and the squashed merge.
+        fork_point = self.get_merge_base(reachable_from_commit_hash, equivalent_to_commit_hash)
+        if not fork_point:
+            return False
+
+        equivalent_changeset = self._popen_git(
+            "diff",
+            fork_point,
+            equivalent_to_commit_hash
+        ).stdout
+        equivalent_patch_id = self.get_patch_id(equivalent_changeset)
+
+        if equivalent_patch_id is None and equivalent_changeset.strip() == '':
+            # Empty changeset means the branches are identical, so the tree is equivalent.
+            self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = True
+            return True
+
+        for logentry in self.get_commits_between(fork_point, reachable_from_commit_hash):
+            changeset = self._popen_git(
```

Comment:
Then no loop necessary 🎉 so still `O(N_managed_branches)` git invocations.

Reply:
But as for now, let's punt flag/config key, I can always add it later

Reply:
Ha! even better, git cherry (https://git-scm.com/docs/git-cherry), seems to be created for this particular purpose

Reply:
(If only you can align the arguments required by git cherry right with what's required here)

Reply:
I checked out `git cherry`, but it suffers from the same problem, it can only check if a single commit has been applied, not multiple at the same time.

Reply:
Aww indeed, I missed that... from git-cherry manual:

> Determine whether there are commits in <head>..<upstream> that are equivalent to those in the range <limit>..<head>
> Outputs the SHA1 of every commit in <limit>..<head>, prefixed with - for commits that have an equivalent in <upstream>, and + for commits that do not.

There might be a workaround if needed: creating a synthetic squash commit on the go using `git commit-tree`, similarly to what `git machete squash` does. Still, there might still be a problem with aligning args to `git cherry` correctly (even more so).

Let's leave as is for now... I'm gonna test your implementation for some time on my large commercial repo, with `GIT_MACHETE_MEASURE_COMMAND_TIME` env var (undocumented, internal) to see how much time is spent in each underlying `git` command.

Reply:
And then let's decide if this new mode should be kept behind a flag/config key, or maybe it could be seamlessly made a default

Reply:
(https://github.com/VirtusLab/git-machete/pull/1245)